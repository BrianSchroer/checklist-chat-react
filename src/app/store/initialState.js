import * as modalDialogType from '../modalDialogType';

export default {
    rooms: [],
    currentRoomId: null,
    ajaxCallsInProgressCount: 0,
    modalDialog: modalDialogType.NONE
};
