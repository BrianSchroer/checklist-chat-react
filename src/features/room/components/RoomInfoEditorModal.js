import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {loadChatMessagesForRoom} from '../../../features/chat/chatMessageDucks';
import {saveRoomInfo} from '../../../features/room/roomDucks';
import {validate} from '../roomInfoValidator';
import ModalContainer from '../../../components/ModalContainer';
import TextInput from '../../../components/TextInput';

class RoomInfoEditorModal extends React.Component {
    constructor(props, context) {
        super(props, context);

        this.state = {
            room: Object.assign({}, this.props.room),
            isSaving: false,
            isDeleting: false,
            isDirty: false,
            errors: {}
        };

        this.onChange = this.onChange.bind(this);
        this.onSave = this.onSave.bind(this);
    }

    onChange(event) {
        const room = this.state.room;
        room[event.target.name] = event.target.value;
        return this.setState({room: room, isDirty: true});
    }

    onSave(event) {
        event.preventDefault();

        const {room, isDirty, isNewRoom} = this.state;
        const {actions, onCloseRequest} = this.props;

        if (!isNewRoom && !isDirty) {
            onCloseRequest(event);
            return;
        }

        const validationResponse = validate(room);

        if (!validationResponse.isValid) {
            this.setState({errors: validationResponse.errors});
            return;
        }

        actions.saveRoomInfo(room);
        onCloseRequest(event);

        if (isNewRoom) {
            //TODO: route to new room?
        } else {
            actions.loadChatMessagesForRoom(room.id);
        }
    }

    render() {
        const {isNewRoom, onCloseRequest} = this.props;
        const {room, errors} = this.state;

        return (
            <ModalContainer
                title={(isNewRoom) ? 'Add Chat Room' : 'Edit Chat Room Info'}
                onCloseRequest={onCloseRequest}>

                <div className="modal-body">
                    <TextInput name="roomName" label="Chat Room Name" value={room.roomName}
                        onChange={this.onChange} error={errors.roomName} />

                    <TextInput name="description" label="Room Description"
                        rows={2} value={room.description}
                        onChange={this.onChange} error={errors.description} />

                    <TextInput name="phoneInfo" label="Phone Info" value={room.phoneInfo}
                        onChange={this.onChange} error={errors.phoneInfo} />
                </div>

                <div className="modal-footer">
                    <input type="button" value="Cancel" className="btn btn-default" onClick={onCloseRequest}/>
                    <input type="submit" value="Save" className="btn btn-primary" onClick={this.onSave}/>
                </div>

            </ModalContainer>
        );
    }
}

RoomInfoEditorModal.propTypes = {
    room: PropTypes.object.isRequired,
    isNewRoom: PropTypes.bool.isRequired,
    onCloseRequest: PropTypes.func.isRequired,
    actions: PropTypes.object.isRequired
};

function emptyRoom() {
    return {id: '', roomName: '', description: '', phoneInfo: ''};
}

function mapStateToProps(state, ownProps) {
    let isNewRoom = true;
    let room = emptyRoom();

    const [roomId] = state.modalDialogRequest.keys;

    if (roomId) {
        isNewRoom = false;
        const rooms = state.rooms;
        room = rooms.find(room => room.id == roomId);
    }

    const onCloseRequest = ownProps.onCloseRequest;

    return {onCloseRequest, room, isNewRoom};
}

const mapDispatchToProps = (dispatch) => ({
    actions: bindActionCreators(
        {saveRoomInfo, loadChatMessagesForRoom},
        dispatch) });

export default connect(mapStateToProps, mapDispatchToProps)(RoomInfoEditorModal);
