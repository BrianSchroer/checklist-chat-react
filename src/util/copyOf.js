export default original => Object.assign({}, original);
