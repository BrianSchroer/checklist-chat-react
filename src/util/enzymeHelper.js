const enzymeHelper = class {

    static nestedFind(wrapper, selector) {
        const selectors = selector.split('>').map(node => node.trim());
        let parent = wrapper;
        let found = [];

        for (let i = 0; i < selectors.length; i++) {
            found = parent.find(selectors[i]);
            if (!found.length) {
                break;
            }
        }

        return found;
    }

    static findSingle(wrapper, selector) {
        const findArray = enzymeHelper.nestedFind(wrapper, selector);

        if (findArray.length != 1) {
            throw(`expected 1 find for selector "${selector}", but found ${findArray.length}.`);
        }

        return findArray.first();
    }
};

export default enzymeHelper;
