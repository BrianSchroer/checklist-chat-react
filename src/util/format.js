import moment from 'moment';

export default class {
    static time(input = '') {

        const isoDateFormatRegex = /\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}([+-]\d{2}:\d{2})?/g;
        let formatted = input;

        if (input && input.match(isoDateFormatRegex)) {
            formatted = this.dateTimeString(input, 'h:mm a');
        }

        return formatted;
    }

    static dateTimeString(input, formatString) {
        let formatted = input;

        if (input) {
            const testMoment = moment(input);

            if (testMoment.isValid()) {
                const testFormat = testMoment.format(formatString);
                if (testFormat != 'Invalid date') {
                    formatted = testFormat;
                }
            }
        }

        return formatted;
    }
}
