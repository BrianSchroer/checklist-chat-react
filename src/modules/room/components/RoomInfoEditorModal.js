import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import {uiHelper} from '../../../util';
import {saveRoomInfo} from '../../../modules/room/roomDucks';
import {validate} from '../roomInfoValidator';
import {Modal, TextInput} from '../../../modules/core';

export class RoomInfoEditorModal extends React.Component {
    static get propTypes() {
        return {
            room: PropTypes.object.isRequired,
            isNewRoom: PropTypes.bool.isRequired,
            shouldFocus: PropTypes.bool.isRequired,
            onCloseRequest: PropTypes.func.isRequired,
            actions: PropTypes.object.isRequired
        };
    }

   constructor(props, context) {
        super(props, context);

        this.state = {
            room: Object.assign({}, this.props.room),
            isSaving: false,
            isDeleting: false,
            isDirty: false,
            errors: {}
        };

        this.onChange = this.onChange.bind(this);
        this.onSave = this.onSave.bind(this);
    }

    componentDidMount() {
        if (this.props.shouldFocus) {
            uiHelper.afterRenderIsComplete(() => {
                uiHelper.setFocusToFirstInputInForm('roomInfoEditorModalForm');
            });
        }
    }

    onChange(event) {
        const room = this.state.room;
        room[event.target.name] = event.target.value;
        return this.setState({room: room, isDirty: true});
    }

    onSave(event) {
        event.preventDefault();

        const {room, isDirty, isNewRoom} = this.state;
        const {actions, onCloseRequest} = this.props;

        if (!isNewRoom && !isDirty) {
            onCloseRequest(event);
            return;
        }

        const validationResponse = validate(room);

        if (!validationResponse.isValid) {
            this.setState({errors: validationResponse.errors});
            return;
        }

        actions.saveRoomInfo(room);
        onCloseRequest(event);
    }

    render() {
        const {isNewRoom, onCloseRequest} = this.props;
        const {room, errors} = this.state;

        const buttons = (
            <div>
                <input type="button" value="Cancel" className="btn btn-default" onClick={onCloseRequest}/>
                <input type="submit" value="Save" className="btn btn-primary" onClick={this.onSave}/>
            </div>
        );

        return (
            <Modal
                formId="roomInfoEditorModalForm"
                title={(isNewRoom) ? 'Add Chat Room' : 'Edit Chat Room Info'}
                onCloseRequest={onCloseRequest}
                buttons={buttons}>

                <div>
                    <TextInput name="roomName" label="Chat Room Name" value={room.roomName}
                        onChange={this.onChange} error={errors.roomName} />

                    <TextInput name="description" label="Room Description"
                        rows={2} value={room.description}
                        onChange={this.onChange} error={errors.description} />

                    <TextInput name="phoneInfo" label="Phone Info" value={room.phoneInfo}
                        onChange={this.onChange} error={errors.phoneInfo} />
                </div>

            </Modal>
        );
    }
}

function emptyRoom() {
    return {id: '', roomName: '', description: '', phoneInfo: ''};
}

export function mapStateToProps(state, ownProps) {
    let isNewRoom = true;
    let room = emptyRoom();

    const [roomId] = state.modalDialogRequest.keys;

    if (roomId) {
        isNewRoom = false;
        const rooms = state.rooms;
        room = rooms.find(room => room.id == roomId);
    }

    const onCloseRequest = ownProps.onCloseRequest;

    const shouldFocus = (state.shouldFocus == undefined) ? true : state.shouldFocus;

    return {onCloseRequest, room, isNewRoom, shouldFocus};
}

const mapDispatchToProps = (dispatch) => ({
    actions: bindActionCreators({saveRoomInfo}, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(RoomInfoEditorModal);
