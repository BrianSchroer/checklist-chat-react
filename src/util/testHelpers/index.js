import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
