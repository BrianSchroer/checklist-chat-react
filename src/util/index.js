import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
