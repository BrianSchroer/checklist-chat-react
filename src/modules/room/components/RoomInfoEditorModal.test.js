import React from 'react';
import { snapshotHelper } from '../../../util/testHelpers';
import initialState from '../../../app/store/initialState';
import { RoomInfoEditorModal, mapStateToProps } from './RoomInfoEditorModal';

const defaultRoom = {
  id: 292,
  roomName: 'TestName',
  description: 'TestDescription',
  phoneInfo: 'TestPhone'
};

const emptyRoom = {
  id: '',
  roomName: '',
  description: '',
  phoneInfo: ''
};

const defaultProps = {
  room: defaultRoom,
  isNewRoom: false,
  shouldFocus: false,
  onCloseRequest: () => {},
  actions: {
    saveRoomInfo: () => {}
  }
};

const defaultOwnProps = {
  onCloseRequest: () => {}
};

function overrideProps(propOverrides) {
  return Object.assign({}, defaultProps, propOverrides);
}

function assertSnapshotMatch(propOverrides = {}) {
  snapshotHelper.assertMatch(
    <RoomInfoEditorModal {...overrideProps(propOverrides)} />
  );
}

function newRoomInfoEditorModal() {
  const props = { room: defaultRoom };
  return new RoomInfoEditorModal(props);
}

function callMapStateToProps(stateOverrides, ownPropsOverrides) {
  const state = Object.assign({}, initialState, stateOverrides || {});
  const ownProps = Object.assign({}, defaultOwnProps, ownPropsOverrides || {});

  return mapStateToProps(state, ownProps);
}

describe('RoomInfoEditorModal', () => {
  describe('mapStateToProps', () => {
    it('should return expected props', () => {
      const props = callMapStateToProps();
      expect(props.onCloseRequest).toBe(defaultOwnProps.onCloseRequest);
    });

    it('should set expected room properties when modalDialogRequest.key exists', () => {
      const testRoom = { id: 123 };

      const stateOverrides = {
        rooms: [{ id: 1 }, { id: testRoom.id }, { id: 999 }],
        modalDialogRequest: { keys: [testRoom.id] }
      };

      const props = callMapStateToProps(stateOverrides);

      expect(props.isNewRoom).toBe(false);
      expect(props.room).toBe(stateOverrides.rooms[1]);
    });

    it('should set expected room properties when modalDialogRequest.key does not exist', () => {
      const stateOverrides = {
        rooms: [],
        modalDialogRequest: { keys: [] }
      };

      const props = callMapStateToProps(stateOverrides);

      expect(props.isNewRoom).toBe(true);
      expect(props.room.id).toBe('');
    });
  });

  it('constructor should set expected component state', () => {
    const state = newRoomInfoEditorModal().state;
    expect(state.room.id).toBe(defaultRoom.id);
    expect(state.isSaving).toBe(false, 'isSaving should be false');
    expect(state.isDeleting).toBe(false, 'isDeleting should be false');
    expect(state.isDirty).toBe(false, 'isDirty should be false');
    expect(Object.keys(state.errors).length).toBe(
      0,
      'errors should be empty object'
    );
  });

  it('should render correctly for new room', () => {
    assertSnapshotMatch({ isNewRoom: true, room: emptyRoom });
  });

  it('should render correctly for existing room', () => {
    assertSnapshotMatch();
  });
});
