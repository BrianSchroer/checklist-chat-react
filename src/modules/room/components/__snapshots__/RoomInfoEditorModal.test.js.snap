// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomInfoEditorModal should render correctly for existing room 1`] = `
<div
  className="checklist-chat-modal-backdrop"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog checklist-chat-modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="modal-header bg-primary text-white"
        >
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
          <h4
            className="modal-title display-linebreaks"
          >
            Edit Chat Room Info
          </h4>
        </div>
        <div
          className="modal-body"
        >
          <div>
            <div
              className="form-group"
            >
              <label
                htmlFor="roomName"
              >
                Chat Room Name
              </label>
              <div
                className="field"
              >
                <input
                  className="form-control"
                  name="roomName"
                  onChange={[Function]}
                  placeholder={undefined}
                  type="text"
                  value="TestName"
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="description"
              >
                Room Description
              </label>
              <div
                className="field"
              >
                <textarea
                  className="form-control"
                  name="description"
                  onChange={[Function]}
                  placeholder={undefined}
                  rows={2}
                  value="TestDescription"
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="phoneInfo"
              >
                Phone Info
              </label>
              <div
                className="field"
              >
                <input
                  className="form-control"
                  name="phoneInfo"
                  onChange={[Function]}
                  placeholder={undefined}
                  type="text"
                  value="TestPhone"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <div>
            <input
              className="btn btn-default"
              onClick={[Function]}
              type="button"
              value="Cancel"
            />
            <input
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
              value="Save"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`RoomInfoEditorModal should render correctly for new room 1`] = `
<div
  className="checklist-chat-modal-backdrop"
  role="dialog"
  tabIndex="-1"
>
  <div
    className="modal-dialog checklist-chat-modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <form
        onSubmit={[Function]}
      >
        <div
          className="modal-header bg-primary text-white"
        >
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
          <h4
            className="modal-title display-linebreaks"
          >
            Add Chat Room
          </h4>
        </div>
        <div
          className="modal-body"
        >
          <div>
            <div
              className="form-group"
            >
              <label
                htmlFor="roomName"
              >
                Chat Room Name
              </label>
              <div
                className="field"
              >
                <input
                  className="form-control"
                  name="roomName"
                  onChange={[Function]}
                  placeholder={undefined}
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="description"
              >
                Room Description
              </label>
              <div
                className="field"
              >
                <textarea
                  className="form-control"
                  name="description"
                  onChange={[Function]}
                  placeholder={undefined}
                  rows={2}
                  value=""
                />
              </div>
            </div>
            <div
              className="form-group"
            >
              <label
                htmlFor="phoneInfo"
              >
                Phone Info
              </label>
              <div
                className="field"
              >
                <input
                  className="form-control"
                  name="phoneInfo"
                  onChange={[Function]}
                  placeholder={undefined}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          className="modal-footer"
        >
          <div>
            <input
              className="btn btn-default"
              onClick={[Function]}
              type="button"
              value="Cancel"
            />
            <input
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
              value="Save"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
